<a href="https://colab.research.google.com/github/yandexdataschool/MLatImperial2020/blob/master/02_lab/kaggle_lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Set up variables and download data

Register on [kaggle](https://www.kaggle.com) and accept the [competition](https://www.kaggle.com/t/f62a416111034502a0090ea0c6adcf18) rules.

Go to My Account and under API section click **create new API Token**.
Download created kaggle.json

Upload this file to your google drive root folder.

Now execute the following magic. - It installs kaggle, mounts google drive and downloads data from competition to you drive.

In [0]:
from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


In [0]:
!mkdir /root/.kaggle
!cp /content/gdrive/My\ Drive/kaggle.json /root/.kaggle/
!chmod 600 /root/.kaggle/kaggle.json
!ls -l /root/.kaggle

total 4
-rw------- 1 root root 64 Jan 28 12:02 kaggle.json


In [0]:
DATA_PATH = "/content/gdrive/My Drive/mlimperial2020-1"

In [0]:
ls /content/gdrive/My\ Drive/mlimperial2020-1

macro.csv.zip  X_test.csv.zip  X_train.csv.zip


In [0]:
#!kaggle config set -n path -v /content
!kaggle competitions download -c mlimperial2020-1 -p '/content/gdrive/My Drive/mlimperial2020-1'

X_train.csv.zip: Skipping, found more recently modified local copy (use --force to force download)
X_test.csv.zip: Skipping, found more recently modified local copy (use --force to force download)
macro.csv.zip: Skipping, found more recently modified local copy (use --force to force download)


# https://www.kaggle.com/t/f62a416111034502a0090ea0c6adcf18

### Metric

For regression task we can use the most common Mean Squared Error(MSE). However, sometimes its better to use logarithmic error. In this challenge, we will use RMSLE - root mean square logarithmic error:

$$
RMSLE = \sqrt{\frac{1}{N} \sum_{i=1}^{N} [\log(y_i + 1) - \log(p_i + 1)]^2},
$$

where $y_i$ is true value and $p_i$ is a predicted value.

# Grading

Your task is to try as many techniques that you have learned this week as possible.


The outcome of your work should be a small table with results, i.e Method - parameters tuned with CV - score + features created on top of exiting ones. The table should be accompanied by a small report of your workflow and reasoning. Also, you need to send the code.


The archive with the files should be sent to mlaticl2020@yandex.ru with the topic: Surname_name_kaggle_1

### The total amount of points is 10. You will get additional points based on your final ranking

- 1 Point. Find correlated features in the train.csv and macro.csv. Try to run linear regression when you remove this features. What do you observe?
- 1 Point. Try to run various linear methods such Ridge, ElasticNet and more. Grid search parameters.
- 1 Point. Work with missed values. Try to impute them, remove them, or someshow other use clustering technic to fill in missed values with the best value.
- 1 Point. Work with categorial features. Find them. Try to one-hot encode them. Does this improves your score? Try to use standard techniques to work with them, such as counting them, calculating frequency, inverted frequency.
- 1 Point. Work with the datastamps. What information can you extract from them? Can you come up with some date based features?
- 1 Point. Try to find badly defined features and outliers in the dataset. Remove them. Did it help?
- 1 Point. Try using PCA/SVD. Is it usefull? Why?
- 1 Point. Create your own features and explain, why did you decide to create those particular ones. Did they make score better?
- 1 Point. Apply decision tree, random forest, boosting based algos. Grid search parameters.
- 1 Point. Estimate feature importances. Try to remove bad features. Which difference did you notice in comparison when you remove correlated features? 
- 1 Point. Use stacking and blending of the models trained above? Does it improve your score?

## Bonus

Beat medium baseline and we will give you +3 points :)

# Baseline

In [0]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_log_error,mean_squared_error
import os

In [0]:
data = pd.read_csv(os.path.join(DATA_PATH, 'X_train.csv.zip'), parse_dates=['timestamp'])
test = pd.read_csv(os.path.join(DATA_PATH, 'X_test.csv.zip'), parse_dates=['timestamp'])
macro = pd.read_csv(os.path.join(DATA_PATH, "macro.csv.zip"), parse_dates=['timestamp'])

In [0]:
data.shape, test.shape, macro.shape

((21329, 292), (9142, 291), (2484, 100))

In [0]:
data.head()

id  timestamp  full_sq  ...  sport_count_5000  market_count_5000  price_doc
0  11180 2013-10-11       51  ...                48                 11    9300000
1  23071 2014-09-30       48  ...               114                 13    5277083
2  18942 2014-05-27       77  ...                33                  8   12500000
3  24805 2014-11-17      109  ...                16                  0   13050000
4  14778 2014-02-10       33  ...                90                  6    5500000

[5 rows x 292 columns]

In [0]:
macro.head()

timestamp  oil_urals  ...  apartment_build  apartment_fund_sqm
0 2010-01-01       76.1  ...          22825.0                 NaN
1 2010-01-02       76.1  ...          22825.0                 NaN
2 2010-01-03       76.1  ...          22825.0                 NaN
3 2010-01-04       76.1  ...          22825.0                 NaN
4 2010-01-05       76.1  ...          22825.0                 NaN

[5 rows x 100 columns]

As you can see, the timestamp is important here, because it will define the various variables, that change with time, for example, gdp or mortgage rate. Lets, for example, merge train, test,the data on the timestamp.

In [0]:
y_train = data["price_doc"]

data.drop(['id', 'price_doc'], axis=1, inplace=True)

# num_train = len(X_train)
# X_all = pd.concat([X_train, X_test])


X_all = pd.merge_ordered(data, macro, on='timestamp', how='left')


In [0]:
X_all.head()

timestamp  full_sq  ...  apartment_build  apartment_fund_sqm
0 2011-08-20       43  ...          23587.0            230310.0
1 2011-08-23       34  ...          23587.0            230310.0
2 2011-08-27       43  ...          23587.0            230310.0
3 2011-09-01       89  ...          23587.0            230310.0
4 2011-09-08       25  ...          23587.0            230310.0

[5 rows x 389 columns]

A small hint - do we really need all 389 columns? What is the distribution of the predicted data?

For now, lets split the training set to train/test and fit simpliest linear model on top of it. But before that we must get rid of NaNs!, beacause not algorithms can deal with them.

In [0]:
X_all.fillna(0, inplace=True)

In [0]:
training_ind, validation_ind = train_test_split(range(len(X_all)), random_state=11, train_size=0.10)

In [0]:
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.model_selection import GridSearchCV

In [0]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_all.iloc[training_ind])

TypeError: ignored

# Error! Why?

Because we have there are exists categorial values in the table. For now, we will just drop them, but you should not! They might be important for the prediction result.

In [0]:
df_numeric = X_all.select_dtypes(exclude=['object'])
df_numeric.drop(["timestamp"], inplace=True, axis=1)

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:4117: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [0]:
df_numeric.head()

full_sq  life_sq  ...  apartment_build  apartment_fund_sqm
0       43     27.0  ...          23587.0            230310.0
1       34     19.0  ...          23587.0            230310.0
2       43     29.0  ...          23587.0            230310.0
3       89     50.0  ...          23587.0            230310.0
4       25     14.0  ...          23587.0            230310.0

[5 rows x 370 columns]

In [0]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(df_numeric.iloc[training_ind])

In [0]:
predictor = Ridge()
predictor.fit(X_train, np.log1p(y_train[training_ind]))

X_test = scaler.transform(df_numeric.iloc[validation_ind])
mean_squared_error(predictor.predict(X_test), np.log1p(y_train[validation_ind]), squared=False)

0.6451176136016002

In [0]:
%%time

param_grid = {'alpha': [0.001, 0.01, 0.1, 1, 10]}

gscv = GridSearchCV(predictor, param_grid, scoring='neg_root_mean_squared_error', cv=3, n_jobs=-1, verbose=1)
gscv.fit(X_train, np.log1p(y_train[training_ind]))

Fitting 3 folds for each of 5 candidates, totalling 15 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.


CPU times: user 114 ms, sys: 55.6 ms, total: 169 ms
Wall time: 2.09 s


[Parallel(n_jobs=-1)]: Done  15 out of  15 | elapsed:    2.0s finished


In [0]:
gscv.cv_results_

{'mean_fit_time': array([0.04698173, 0.03693652, 0.03917631, 0.03910669, 0.03382055]),
 'mean_score_time': array([0.00165669, 0.00164835, 0.00167902, 0.00165915, 0.00143957]),
 'mean_test_score': array([-0.6880552 , -0.68577622, -0.68010392, -0.66494362, -0.64622507]),
 'param_alpha': masked_array(data=[0.001, 0.01, 0.1, 1, 10],
              mask=[False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'alpha': 0.001},
  {'alpha': 0.01},
  {'alpha': 0.1},
  {'alpha': 1},
  {'alpha': 10}],
 'rank_test_score': array([5, 4, 3, 2, 1], dtype=int32),
 'split0_test_score': array([-0.74052183, -0.73928826, -0.736059  , -0.72651571, -0.71363541]),
 'split1_test_score': array([-0.66967212, -0.6648102 , -0.65658956, -0.63739805, -0.61208545]),
 'split2_test_score': array([-0.65397166, -0.6532302 , -0.6476632 , -0.63091709, -0.61295436]),
 'std_fit_time': array([0.00700515, 0.00209948, 0.00153328, 0.00128192, 0.00547819]),
 'std_score_time': array([6.9

## Now we can fit on all the data and make a prediction

Refit model with best parameters

In [0]:
predictor = Ridge(alpha=1)
predictor.fit(X_train, np.log1p(y_train[training_ind]))

X_test = scaler.transform(df_numeric.iloc[validation_ind])
mean_squared_error(predictor.predict(X_test), np.log1p(y_train[validation_ind]), squared=False)

0.6451176136016002

# Make predictions on the test set



In [0]:
test.head()

id  timestamp  ...  sport_count_5000  market_count_5000
0   0 2013-06-25  ...                12                  1
1   1 2014-11-20  ...                11                  1
2   2 2014-12-29  ...                49                  1
3   3 2012-10-03  ...                 8                  1
4   4 2014-11-11  ...               209                 10

[5 rows x 291 columns]

In [0]:
pred_ids = test['id']
test.drop(['id'], axis=1, inplace=True)
X_predict = pd.merge_ordered(test, macro, on='timestamp', how='left')
X_predict.fillna(0, inplace=True)
X_predict = X_predict[df_numeric.columns]

In [0]:
predictions = np.expm1(predictor.predict(scaler.transform(X_predict)))
predictions = pd.DataFrame(predictions, columns=["price_doc"])
predictions = pd.concat([pred_ids, predictions], axis=1)

In [0]:
predictions.to_csv(os.path.join(DATA_PATH, "predictions.csv"), index=False)

In [0]:
!head -n 5 '/content/gdrive/My Drive/mlimperial2020-1/predictions.csv'

id,price_doc
0,5709943.84614118
1,5752432.388775702
2,7190087.855122229
3,7337914.284958483


# Lets use kaggle API again to submit results


In [0]:
!kaggle competitions submit -c mlimperial2020-1 -f "{DATA_PATH}/predictions.csv" -m "Message"

100% 204k/204k [00:04<00:00, 51.3kB/s]
403 - This competition is not currently active.  To test submissions prior to competition launch, use the Sandbox Submissions tool found on the Host tab.
